### ISSUE TITLE
Template source maps cause issues for source-map-explorer

### ISSUE THREAD

**Issue by IgorMinar** [angular/angular#19840]
## Current behavior
Before the recent fix that enabled template source maps, the source-map explorer showed generated factories with full path. After the recent update there are at least two issues with the template source maps:
- [ ] source maps are not prefixed with the path of the original component / template file, causing all component templates to be mixed
- [ ] there is a lot of unaccounted space in the source-map-explorer

<img width="1446" alt="screen shot 2017-10-20 at 3 35 05 pm" src="https://user-images.githubusercontent.com/216296/31844412-62b80c3c-b5ad-11e7-8671-9e1baa98c34e.png">

## Expected behavior
- template source maps should be prefixed with the path of the original component/template
- there shouldn't be any significant unaccounted space

## Environment

<pre><code>
Angular CLI: 1.5.0-rc.2
Node: 6.9.5
OS: darwin x64
Angular: 5.0.0-rc.2
... animations, common, compiler, compiler-cli, core, forms
... http, platform-browser, platform-browser-dynamic
... platform-server, router

@angular/cdk: 2.0.0-beta.12
@angular/cli: 1.5.0-rc.2
@angular/material: 2.0.0-beta.12
@angular/service-worker: 1.0.0-beta.16
@angular-devkit/build-optimizer: 0.0.29
@angular-devkit/core: 0.0.20
@angular-devkit/schematics: 0.0.34
@ngtools/json-schema: 1.1.0
@ngtools/webpack: 1.8.0-rc.2
@schematics/angular: 0.0.48
typescript: 2.5.3
webpack: 3.7.1
</code></pre>

**Comment by IgorMinar** [comment]
I'm marking this as a regression because previously the source-map-explorer info was very useful and now it's no longer useful.

**Comment by IgorMinar** [comment]
Potential workaround would be to allow template source maps to be disabled using an `angularCompilerOptions` flag: #19858

**Comment by tbosch** [comment]
It would be helpful to really debug this before we start adding features that might not be needed in the future. A local hack is the following:

Comment out the line `converter.updateSourceMap(sourceStatments);` in `@angular/compiler-cli/src/transformers/node_emitter.js` .

**Comment by tbosch** [comment]
A good solution could also be that we map all code that does not correspond to a line in the html file to the `.ngfactory.ts` file...

**Comment by IgorMinar** [comment]
Mapping the code to two files seems confusing. If this is a
source-map-explorer issue then we can file an issue against it. @kara could
fix it since she's familiar with their code.

I do think that having an option to disable source map generation from
templates can be useful for debugging. But I can be convinced of otherwise
if the debugging and source-map-explorer experience is great with the
source maps on.

Should we track the problem with paths in a separate issue?

**Comment by tbosch** [comment]
Soon we will inline the generated code into the `.ts` file itself. What would we do then? There won't be any `.ngfactory` files anymore. Where would we then attribute the generated code to?

**Comment by IgorMinar** [comment]
@tbosch that's a fair point. however, source-map-explorer is an essential tool for debugging size regressions. Breaking it is not acceptable. Have you verified that your suggested patch works? I can't test it now, but I did mess with node_emitter.js before you suggested it and tried commenting out those lines and it resulted in internal errors being thrown.

Also keep in mind that when debugging issues, it can use useful to debug raw factories (without source maps), while having source maps for all the hand written code.

**Comment by chuckjaz** [comment]
@IgorMinar Can you be more specific on when you thought it broke? My preliminary investigation shows that this has been broken from the point we switched to the transform based compiler which was fairly early in the 5.0.0 time frame. What version of 5.0.0 are you seeing what you expect?

**Comment by IgorMinar** [comment]
@chuckjaz the last time I know this worked fine was in 5.0.0-beta.3. But I didn't discover this problem until recently because our source maps were messed up for other reasons related to packaging. When those were resolved and we tried to upgrade to rc.2 we discovered that our source maps no longer worked with source-map-explorer because of the template source maps feature.

**Comment by chuckjaz** [comment]
@IgorMinar 

1) The fact that paths are not included in the .html is because resolving the path is a done in `loader.ts` of the `@angular/cli` (https://github.com/angular/angular-cli/blob/master/packages/@ngtools/webpack/src/loader.ts#L578). This line only updates the location of the first source file (which is the `.ts` file) and ignores the other files (like the `.html` files) we added to the source map so the path is getting lost.

2) We had a bug in our source maps where we were producing a mapping for a blank url. I will have  a PR out, shortly, to fix this.

3) The `undefined` regions are not factory file related. They seem to be from other parts of `aio`. I determined this by writing program that used the source explorer algorithm to map every character of the file and produces an index back to the original file.

4) We cannot simply disable Angular's source maps for `.ngfactory.ts` files and just let TypeScript produce a source map as there is no `.ngfactory.ts` file which TypeScript could use. The code is generated in the transformer pipeline which are synthetic nodes with no reference to source code. The only source code we have is the original template which is why we map to that. 

5) For size regression analysis, we recommend not producing source maps from `ngc` at all.
   a) Producing source map for `.ngfactory` files that only contain an `@NgModule` would increase our source map size which is already very large. For instance, our simple example in G3 already is 20 megs of source maps. These addition mapping would only useful in size regression analysis.
    b) For the size regression analysis, it would be more accurate to analyze what generated `.js` file contributed to the bundle instead of the original `.ts` as the source maps produced by TypeScript for `.ts` incompletely map the generated `.js` back to the `.ts` producing misleading results.

6) Producing source maps that are intended to be used for size regressions, instead of debugging, should be an option added to the `ng` which would skip the source maps from TypeScript/ngc.

*Action Items*

* The PR to fix the bug in our source map production should be fixed before 5.0.0 release and I will have a PR ready for tonight.
* I will speak with Hans about fixing the path of additional files in a `.ts` source map.

**Should we hold the release for this?**

I don't believe we need to. Producing the size regression source maps change in `ng` doesn't require changes in Angular and is a non-trivial implementation (probably 2-3 days) and will produce more accurate analysis.

Note: we demonstrated that identity source maps are not a good solution as they don't produce the correct results and confuse other down stream tools. Just removing the source maps (as you tried) requires modifications to `ng` as it assumes that if it asked for a source file it will be produced.

**Comment by tbosch** [comment]
The PR Chuck mentioned is here: https://github.com/angular/angular/pull/19965

**Comment by IgorMinar** [comment]
Thanks Chuck, this is very insightful and explains quite a bit.

So it sounds like the only option for dealing with the undefined regions is
to make it possible for the cli to disable tsc source maps while still
producing the final bundle with a source maps that map back to the tsc
output. Is that right? Is Hans ok with that? If it works for him then I'll
be fine with that solution.

**Comment by chuckjaz** [comment]
@IgorMinar Currently Hans and I are discussing what the options are but we agree that the source maps produced by `ngc` are sufficient for any solution we come up with.

To be clear, "undefined" is coming from something downstream in the tool chain. and are not related to the source maps produced by the template compiler. For example, the following text form the main bundle is marked as "undefined":

```
if(!t||"object"!=typeof t)throw e.logger.log("received invalid data:",t),Error("Invalid data")}).catch(function(n){return 404===n.status?e.getFileNotFoundDoc(t):e.getErrorDoc(t,n)})
```

This section is from `document.service.ts`. We should investigate where this is coming from. I am currently verifying that correct source maps are produced for development mode.

**Comment by IgorMinar** [comment]
with core@5.0.0-rc.5 and cli@1.5.0-rc.6 the source maps look different (but not more useful):

<img width="1612" alt="screen shot 2017-10-30 at 11 39 53 am" src="https://user-images.githubusercontent.com/216296/32166824-610e1072-bd67-11e7-9e83-c0faa69c31fa.png">

**Comment by chuckjaz** [comment]
@IgorMinar: I investigated more with @hansl and we found an issue in `--build-optimizer` that is being addressed by https://github.com/angular/devkit/pull/247.

**Comment by hansl** [comment]
I'm closing this issue as both PRs (https://github.com/angular/devkit/pull/247 and https://github.com/angular/angular-cli/pull/8257) that fix this are in master of CLI and devkit.

**Comment by IgorMinar** [comment]
Nice work! Thanks for getting this done. I'll test the fixes tomorrow.

**Comment by saiprakash1** [comment]
Excellent, need to upgrade my knowledge base and thereby project
